src/chainrules.jl

```diff
@@ -0,0 +1,40 @@
+@non_differentiable validate_woodbury_arguments(A, D, S)
+
+function ChainRulesCore.rrule(
+    ::typeof(*), A::Real, B::WoodburyPDMat{T, TA, TD, TS}
+) where {T, TA, TD, TS}
+    project_A = ProjectTo(A)
+    project_B = ProjectTo(B)
+    function times_pullback(ȳ::AbstractMatrix)
+        Ȳ = unthunk(ȳ)
+        Ā = dot(Ȳ, B)
+        B̄ = A' * Ȳ
+        return (
+           NoTangent(),
+           @thunk(project_A(Ā')),
+           @thunk(project_B(B̄)),
+        )
+    end
+
+    function times_pullback(ȳ::Tangent{<:WoodburyPDMat})
+        Ȳ = unthunk(ȳ)
+        Ā = dot(Ȳ.A * Ȳ.D * Ȳ.A' + Ȳ.S, B)
+        B̄ = Ȳ.A * (A' * Ȳ.D) * Ȳ.A' + A' * Ȳ.S
+        return (
+           NoTangent(),
+           @thunk(project_A(Ā')),
+           @thunk(project_B(B̄)),
+        )
+    end
+    return A * B, times_pullback
+end
+
+function ChainRulesCore.ProjectTo(W::WoodburyPDMat)
+    function dW(W̄)
```

Comment:
Is it not possible to construct `ProjectTo` object? And then make it callable? All the examples in the ChainRulesCore are built this way, see https://github.com/JuliaDiff/ChainRulesCore.jl/blob/master/src/projection.jl.

We have decided against closures because they are not easily extensible. I.e. if someone defines their own type and wants to make sure it can be projected to a Woodbury then they must either pirate or change this function (and introduce a dependency to PDMatsExtras.

Reply:
Oh right. Will change.

Reply:
updated